compiler/parser/RdrHsSyn.hs

```diff
@@ -2274,19 +2303,70 @@ funBind
   :: SrcSpan
   -> Located RdrName
   -> MatchGroup GhcPs (LHsExpr GhcPs)
-  -> P (Maybe (LHsBind GhcPs))
+  -> LHsBind GhcPs
 funBind loc tag mg =
-  return $ Just $ L loc $
+  L loc $
     FunBind { fun_ext = noExt
             , fun_id = tag
             , fun_matches = mg
             , fun_co_fn = WpHole
             , fun_tick = []
             }
 
--- | Put all your cats in a bag.
-bagOfCatMaybes :: [Maybe a] -> Bag a
-bagOfCatMaybes = listToBag . catMaybes
+mkRdrName :: String -> RdrName
+mkRdrName = mkRdrUnqual . mkVarOcc
+
+mkTemplateClassInstanceSigs :: String -> Maybe (LHsType GhcPs) -> [LSig GhcPs]
+mkTemplateClassInstanceSigs templateName mbKeyType =
+  map (uncurry mkSig) $
+    [ ("signatory", mkFunTy templateType parties)
+    , ("observer",  mkFunTy templateType parties)
+    , ("ensure",    mkFunTy templateType (mkQualTypeName "Bool"))
+    , ("agreement", mkFunTy templateType (mkQualTypeName "Text"))
+    , ("create",    mkFunTy templateType (update $ mkParenTy contractId))
+    , ("fetch",     mkFunTy contractId (update templateType))
+    , ("archive",   mkFunTy contractId (update unit))
+    ]
+    ++ keySigs
+  where
+    keySigs = flip (maybe []) mbKeyType $ \keyType ->
+      [ ("hasKey",      hasKeyType)
+      , ("key",         mkFunTy templateType keyType)
+      , ("maintainer",  mkFunTy hasKeyType (mkFunTy keyType parties))
+      , ("fetchByKey",  mkFunTy keyType (update $ pairType contractId templateType))
+      , ("lookupByKey", mkFunTy keyType (update $ mkParenTy $ mkAppTy (mkQualTypeName "Optional") (mkParenTy contractId)))
+      ]
+    mkSig :: String -> LHsType GhcPs -> LSig GhcPs
+    mkSig methodName ty =
+      let fullMethodName = noLoc $ mkRdrName $ prefixTemplateClassMethod $ methodName ++ templateName in
+      noLoc $ ClassOpSig noExt False [fullMethodName] (HsIB noExt ty)
+    mkFunTy funTy argTy = noLoc $ HsFunTy noExt funTy argTy
```

Comment:
There seems to be duplication of these utilities `mkFunTy`, `mkAppTy` (see `mkTemplateChoiceSigs` for example)